File: android/app/src/main/java/com/stablechannels/app/services/SpliceBroadcastChecker.kt
package com.stablechannels.app.services

import android.util.Log
import okhttp3.OkHttpClient
import okhttp3.Request

/** Outcome of asking esplora whether it has ever heard of a txid. */
enum class TxBroadcastStatus { EXISTS, NOT_FOUND, INCONCLUSIVE }

/**
 * Checks whether esplora has ever heard of a txid (broadcast, mempool, or confirmed) — used to
 * tell a genuinely abandoned/never-broadcast splice tx apart from a stale failure event for a
 * splice that did make it on-chain.
 *
 * The result is deliberately tri-state rather than a boolean:
 * - EXISTS: any endpoint returned a successful (2xx) response for the tx.
 * - NOT_FOUND: every reachable endpoint returned an explicit 404 across [retries] consecutive
 *   rounds. A single round of 404s isn't proof — the tx may still be propagating right after
 *   negotiation — so this is only concluded once the same negative result is confirmed on retry.
 * - INCONCLUSIVE: anything else (timeouts, exceptions, 429/5xx, or a mix of endpoint results). We
 *   cannot prove the tx doesn't exist here, so callers must never treat this as a real failure —
 *   the safer failure mode is treating a real failure as a stale replay (recoverable manually)
 *   rather than mislabeling a possibly-real splice as failed.
 */
class SpliceBroadcastChecker(
    private val httpClient: OkHttpClient,
    private val retries: Int = 3,
    private val retryDelayMs: Long = 2_000L,
    private val sleep: (Long) -> Unit = { Thread.sleep(it) },
    private val logWarning: (String) -> Unit = { Log.w("SpliceBroadcastChecker", it) }
) {
    fun checkStatus(txid: String, endpointUrls: List<String>): TxBroadcastStatus {
        val normalizedTxid = txid.substringBefore(":")
        val urls = endpointUrls.filter { it.isNotBlank() }.distinct()
        if (urls.isEmpty()) return TxBroadcastStatus.INCONCLUSIVE

        repeat(retries) { attempt ->
            var allNotFound = true
            var anyReached = false
            for (baseUrl in urls) {
                try {
                    val request = Request.Builder()
                        .url("${baseUrl.trimEnd('/')}/tx/$normalizedTxid/status")
                        .build()
                    httpClient.newCall(request).execute().use { response ->
                        anyReached = true
                        if (response.isSuccessful) return TxBroadcastStatus.EXISTS
                        if (response.code != 404) allNotFound = false
                    }
                } catch (e: Exception) {
                    logWarning("Existence check failed: ${e.message}")
                    allNotFound = false
                }
            }
            if (!anyReached || !allNotFound) return TxBroadcastStatus.INCONCLUSIVE
            if (attempt < retries - 1) sleep(retryDelayMs)
        }
        return TxBroadcastStatus.NOT_FOUND
    }
}

File: ios/StableChannels/StableChannels/Services/SpliceBroadcastChecker.swift
import Foundation
import OSLog

/// Outcome of asking Esplora whether it has ever heard of a txid.
public enum TxBroadcastStatus: Equatable, Sendable {
    case exists
    case notFound
    case inconclusive
}

/// Defines a contract for verifying whether a transaction made it on-chain/mempool.
public protocol SpliceBroadcastChecking: Sendable {
    func checkStatus(txid: String, endpointURLs: [String]) async -> TxBroadcastStatus
}

/// Checks whether Esplora has ever heard of a txid (broadcast, mempool, or confirmed) — used to
/// tell a genuinely abandoned/never-broadcast splice tx apart from a stale failure event for a
/// splice that did make it on-chain.
///
/// The result is tri-state:
/// - `.exists`: Any endpoint returned a successful (2xx) response for the tx.
/// - `.notFound`: Every reachable endpoint returned an explicit 404 across all retries.
/// - `.inconclusive`: Timeouts, 5xx, or network errors where non-existence cannot be proven.
public final class SpliceBroadcastChecker: SpliceBroadcastChecking, Sendable {
    private let urlSession: URLSession
    private let retries: Int
    private let retryDelayNanoseconds: UInt64
    private let sleeper: @Sendable (UInt64) async throws -> Void

    private static let logger = Logger(subsystem: "com.stablechannels.app", category: "SpliceBroadcastChecker")

    public init(
        urlSession: URLSession = .shared,
        retries: Int = 3,
        retryDelaySeconds: TimeInterval = 2.0,
        sleeper: @escaping @Sendable (UInt64) async throws -> Void = { try await Task.sleep(nanoseconds: $0) }
    ) {
        self.urlSession = urlSession
        self.retries = retries
        self.retryDelayNanoseconds = UInt64(retryDelaySeconds * 1_000_000_000)
        self.sleeper = sleeper
    }

    public func checkStatus(txid: String, endpointURLs: [String]) async -> TxBroadcastStatus {
        let normalizedTxid = txid.components(separatedBy: ":").first ?? txid
        let urls = endpointURLs.filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
        if urls.isEmpty { return .inconclusive }

        for attempt in 0..<retries {
            var allNotFound = true
            var anyReached = false

            for baseURL in urls {
                let trimmedBase = baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
                guard let url = URL(string: "\(trimmedBase)/tx/\(normalizedTxid)/status") else {
                    continue
                }

                var request = URLRequest(url: url)
                request.httpMethod = "GET"
                request.timeoutInterval = 5.0

                do {
                    let (_, response) = try await urlSession.data(for: request)
                    guard let http = response as? HTTPURLResponse else {
                        allNotFound = false
                        continue
                    }
                    anyReached = true
                    if (200...299).contains(http.statusCode) {
                        return .exists
                    }
                    if http.statusCode != 404 {
                        allNotFound = false
                    }
                } catch {
                    Self.logger
                        .warning(
                            "Existence check for \(normalizedTxid) failed on \(baseURL): \(error.localizedDescription)"
                        )
                    allNotFound = false
                }
            }

            if !anyReached || !allNotFound {
                return .inconclusive
            }

            if attempt < retries - 1 {
                do {
                    try await sleeper(retryDelayNanoseconds)
                } catch {
                    return .inconclusive
                }
            }
        }

        return .notFound
    }
}
